internal/lang/evalchecks/eval_lifecycle_enabled.go

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) The OpenTofu Authors
+// SPDX-License-Identifier: MPL-2.0
+// Copyright (c) 2023 HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+package evalchecks
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/hcl/v2"
+	"github.com/opentofu/opentofu/internal/addrs"
+	"github.com/opentofu/opentofu/internal/lang"
+	"github.com/opentofu/opentofu/internal/lang/marks"
+	"github.com/opentofu/opentofu/internal/tfdiags"
+	"github.com/zclconf/go-cty/cty"
+	"github.com/zclconf/go-cty/cty/convert"
+)
+
+// EvaluateEnabledExpression evaluates an expression assigned to an "enabled"
+// meta-argument and returns either its boolean result or errors describing
+// why such a result cannot be evaluated.
+func EvaluateEnabledExpression(expr hcl.Expression, hclCtxFunc ContextFunc) (bool, tfdiags.Diagnostics) {
+	var diags tfdiags.Diagnostics
+
+	if expr == nil {
+		panic("EvaluateEnabledExpression should be called with a non-nil expr. This is a programming error in OpenTofu.")
+	}
+
+	refs, refsDiags := lang.ReferencesInExpr(addrs.ParseRef, expr)
+	diags = diags.Append(refsDiags)
+	var hclCtx *hcl.EvalContext
+	hclCtx, refsDiags = hclCtxFunc(refs)
+	diags = diags.Append(refsDiags)
+	if diags.HasErrors() { // Can't continue if we don't even have a valid scope
+		return false, diags
+	}
+
+	rawEnabledVal, enabledDiags := expr.Value(hclCtx)
+	diags = diags.Append(enabledDiags)
+
+	if rawEnabledVal.HasMark(marks.Sensitive) {
+		diags = diags.Append(&hcl.Diagnostic{
+			Severity:    hcl.DiagError,
+			Summary:     "Invalid enabled argument",
+			Detail:      "Sensitive values, or values derived from sensitive values, cannot be used in \"enabled\" arguments. If used, the sensitive value would be exposed by whether an instance is present.",
+			Subject:     expr.Range().Ptr(),
+			Expression:  expr,
+			EvalContext: hclCtx,
+			Extra:       DiagnosticCausedByConfidentialValues(true),
+		})
+	}
+
+	enabledVal, err := convert.Convert(rawEnabledVal, cty.Bool)
+	if err != nil {
+		diags = diags.Append(&hcl.Diagnostic{
+			Severity:    hcl.DiagError,
+			Summary:     "Invalid enabled argument",
+			Detail:      fmt.Sprintf(`The given "enabled" argument value is unsuitable: %s.`, err),
+			Subject:     expr.Range().Ptr(),
+			Expression:  expr,
+			EvalContext: hclCtx,
+		})
+	}
```

Comment:
Shouldn't this be moved down, before returning the value from `enabledVal`, after the next `if diags.HasErrors()`? AFAIU, there should be no reason to convert the evaluated value if it's not passing the validations.